scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamOp.scala

```diff
@@ -85,21 +101,113 @@ object BeamOp extends Serializable {
     }
   }
 
-  final case class TransformBeamOp[A, B](source: BeamOp[A], f: PTransform[PCollection[A], PCollection[B]], kryoCoder: KryoCoder) extends BeamOp[B] {
+  final case class TransformBeamOp[A, B](
+    source: BeamOp[A],
+    f: PTransform[PCollection[A], PCollection[B]],
+    kryoCoder: KryoCoder
+  ) extends BeamOp[B] {
     def run(pipeline: Pipeline): PCollection[B] = {
       source.run(pipeline).asInstanceOf[PCollection[A]].apply(f).setCoder(kryoCoder)
     }
   }
 
+  final case class HashJoinOp[K, V, U, W](
+    left: BeamOp[(K, V)],
+    right: BeamOp[(K, U)], joiner: (K, V, Iterable[U]) => Iterator[W]
+  )(implicit kryoCoder: KryoCoder, ordK: Ordering[K]) extends BeamOp[(K, W)] {
+    override def run(pipeline: Pipeline): PCollection[_ <: (K, W)] = {
+      val leftPCollection = left.run(pipeline)
+      val keyCoder: Coder[K] = OrderedSerializationCoder.apply(ordK, kryoCoder)
+      val rightPCollection: PCollection[(K, U)] = widenPCollection(right.run(pipeline))
+
+      val rightPCollectionView = rightPCollection
+        .apply(TupleToKV[K, U](keyCoder, kryoCoder))
+        .apply(GroupByKey.create[K, U]()).setCoder(KvCoder.of(keyCoder, kryoCoder))
+        .apply(View.asMap[K, java.lang.Iterable[U]]())
+
+      leftPCollection
+        .apply(
+          ParDo
+            .of(HashJoinFn[K, V, U, W](joiner, rightPCollectionView))
+            .withSideInputs(rightPCollectionView)
+        )
+        .setCoder(TupleCoder(keyCoder, kryoCoder))
+    }
+  }
+
+  final case class CoGroupedOp[K, V](
+    cg: CoGrouped[K, V],
+    inputOps: Seq[BeamOp[(K, Any)]]
+  )(implicit kryoCoder: KryoCoder) extends BeamOp[(K, V)] {
+    override def run(pipeline: Pipeline): PCollection[_ <: (K, V)] = {
+      val inputOpsSize = inputOps.size
+      val keyCoder: Coder[K] = OrderedSerializationCoder.apply(cg.keyOrdering, kryoCoder)
+      val pcols = inputOps.map { inputOp =>
+        val pCol: PCollection[(K, Any)] = widenPCollection(inputOp.op.run(pipeline))
+        pCol.apply(TupleToKV[K, Any](keyCoder, kryoCoder))
+      }
+
+      val tupleTags = (1 to inputOpsSize).map(idx => new TupleTag[Any](idx.toString))
+
+      val unionCoder = UnionCoder.of(
+        (1 to inputOpsSize).map(_ => kryoCoder.asInstanceOf[Coder[_]]).asJava
+      )
+
+      val coGroupResultCoder: CoGbkResult.CoGbkResultCoder = CoGbkResult.CoGbkResultCoder.of(
+        CoGbkResultSchema.of(tupleTags.map(_.asInstanceOf[TupleTag[_]]).asJava),
+        unionCoder
+      )
+
+      val keyedPCollectionTuple: KeyedPCollectionTuple[K] = pcols
+        .zip(tupleTags)
+        .foldLeft(
+          KeyedPCollectionTuple.empty[K](pipeline)
+        )((keyed, colWithTag) => keyed.and[Any](colWithTag._2, colWithTag._1))
+
+      keyedPCollectionTuple
+        .apply(CoGroupByKey.create()).setCoder(KvCoder.of(kryoCoder, coGroupResultCoder))
+        .apply(ParDo.of(new CoGroupDoFn[K, V](cg, tupleTags))).setCoder(kryoCoder)
```

Comment:
can't we use `.setCoder(KvCoder.of(keyCoder, kryoCoder))` here? if not, can we comment why?